File: kotlin/provenance-mark/src/main/kotlin/com/blockchaincommons/provenancemark/ChaCha20.kt
package com.blockchaincommons.provenancemark

internal class ChaCha20(key: ByteArray, nonce: ByteArray) {
    private val state = UIntArray(16)
    private val keystream = ByteArray(64)
    private var position = 64

    init {
        require(key.size == 32) { "ChaCha20 key must be 32 bytes" }
        require(nonce.size == 12) { "ChaCha20 nonce must be 12 bytes" }

        state[0] = 0x61707865u
        state[1] = 0x3320646eu
        state[2] = 0x79622d32u
        state[3] = 0x6b206574u

        for (i in 0 until 8) {
            val o = i * 4
            state[4 + i] =
                key[o].toUByte().toUInt() or
                    (key[o + 1].toUByte().toUInt() shl 8) or
                    (key[o + 2].toUByte().toUInt() shl 16) or
                    (key[o + 3].toUByte().toUInt() shl 24)
        }

        state[12] = 0u

        for (i in 0 until 3) {
            val o = i * 4
            state[13 + i] =
                nonce[o].toUByte().toUInt() or
                    (nonce[o + 1].toUByte().toUInt() shl 8) or
                    (nonce[o + 2].toUByte().toUInt() shl 16) or
                    (nonce[o + 3].toUByte().toUInt() shl 24)
        }
    }

    fun process(data: ByteArray): ByteArray {
        val out = data.copyOf()
        processInPlace(out)
        return out
    }

    fun processInPlace(data: ByteArray) {
        for (i in data.indices) {
            if (position >= 64) {
                generateBlock()
                position = 0
            }
            data[i] = (data[i].toInt() xor keystream[position].toInt()).toByte()
            position += 1
        }
    }

    private fun generateBlock() {
        val working = state.copyOf()

        repeat(10) {
            quarterRound(working, 0, 4, 8, 12)
            quarterRound(working, 1, 5, 9, 13)
            quarterRound(working, 2, 6, 10, 14)
            quarterRound(working, 3, 7, 11, 15)

            quarterRound(working, 0, 5, 10, 15)
            quarterRound(working, 1, 6, 11, 12)
            quarterRound(working, 2, 7, 8, 13)
            quarterRound(working, 3, 4, 9, 14)
        }

        for (i in 0 until 16) {
            val value = working[i] + state[i]
            val o = i * 4
            keystream[o] = value.toByte()
            keystream[o + 1] = (value shr 8).toByte()
            keystream[o + 2] = (value shr 16).toByte()
            keystream[o + 3] = (value shr 24).toByte()
        }

        state[12] = state[12] + 1u
    }

    private fun quarterRound(s: UIntArray, a: Int, b: Int, c: Int, d: Int) {
        s[a] += s[b]; s[d] = (s[d] xor s[a]).rotateLeft(16)
        s[c] += s[d]; s[b] = (s[b] xor s[c]).rotateLeft(12)
        s[a] += s[b]; s[d] = (s[d] xor s[a]).rotateLeft(8)
        s[c] += s[d]; s[b] = (s[b] xor s[c]).rotateLeft(7)
    }
}

File: csharp/ProvenanceMark/ProvenanceMark/ChaCha20.cs
using System.Buffers.Binary;
using System.Numerics;

namespace BlockchainCommons.ProvenanceMark;

internal sealed class ChaCha20
{
    private readonly uint[] _state = new uint[16];
    private readonly byte[] _keystream = new byte[64];
    private int _position = 64;

    public ChaCha20(ReadOnlySpan<byte> key, ReadOnlySpan<byte> nonce)
    {
        if (key.Length != 32)
        {
            throw new ArgumentException("ChaCha20 key must be 32 bytes", nameof(key));
        }

        if (nonce.Length != 12)
        {
            throw new ArgumentException("ChaCha20 nonce must be 12 bytes", nameof(nonce));
        }

        _state[0] = 0x61707865;
        _state[1] = 0x3320646e;
        _state[2] = 0x79622d32;
        _state[3] = 0x6b206574;

        for (var index = 0; index < 8; index++)
        {
            _state[4 + index] = BinaryPrimitives.ReadUInt32LittleEndian(key[(index * 4)..]);
        }

        _state[12] = 0;

        for (var index = 0; index < 3; index++)
        {
            _state[13 + index] = BinaryPrimitives.ReadUInt32LittleEndian(nonce[(index * 4)..]);
        }
    }

    public byte[] Process(ReadOnlySpan<byte> data)
    {
        var output = data.ToArray();
        ProcessInPlace(output);
        return output;
    }

    public void ProcessInPlace(Span<byte> data)
    {
        for (var index = 0; index < data.Length; index++)
        {
            if (_position >= 64)
            {
                GenerateBlock();
                _position = 0;
            }

            data[index] = (byte)(data[index] ^ _keystream[_position]);
            _position += 1;
        }
    }

    private void GenerateBlock()
    {
        var working = (uint[])_state.Clone();

        for (var round = 0; round < 10; round++)
        {
            QuarterRound(working, 0, 4, 8, 12);
            QuarterRound(working, 1, 5, 9, 13);
            QuarterRound(working, 2, 6, 10, 14);
            QuarterRound(working, 3, 7, 11, 15);

            QuarterRound(working, 0, 5, 10, 15);
            QuarterRound(working, 1, 6, 11, 12);
            QuarterRound(working, 2, 7, 8, 13);
            QuarterRound(working, 3, 4, 9, 14);
        }

        for (var index = 0; index < 16; index++)
        {
            var value = unchecked(working[index] + _state[index]);
            BinaryPrimitives.WriteUInt32LittleEndian(_keystream.AsSpan(index * 4, 4), value);
        }

        _state[12] = unchecked(_state[12] + 1);
    }

    private static void QuarterRound(uint[] state, int a, int b, int c, int d)
    {
        state[a] = unchecked(state[a] + state[b]);
        state[d] = BitOperations.RotateLeft(state[d] ^ state[a], 16);
        state[c] = unchecked(state[c] + state[d]);
        state[b] = BitOperations.RotateLeft(state[b] ^ state[c], 12);
        state[a] = unchecked(state[a] + state[b]);
        state[d] = BitOperations.RotateLeft(state[d] ^ state[a], 8);
        state[c] = unchecked(state[c] + state[d]);
        state[b] = BitOperations.RotateLeft(state[b] ^ state[c], 7);
    }
}
